File: practice/mathematics/number_theory/strange_numbers/strange_numbers.py
#!/usr/bin/env python

from math import log10

def num_digits(i_value):
    if i_value == 0:
        return 1
    return int(log10(i_value)) + 1

STRANGE_VALS = set(range(10))
curr = range(1, 10)
while curr and min(curr) < 10 ** 18:
    new_curr = set([])
    for VAL in curr:
        n = num_digits(VAL)
        d = VAL * n
        if d != VAL and num_digits(d) == n:
            new_curr.add(d)
        d = VAL * (n + 1)
        if num_digits(d) == n + 1:
            new_curr.add(d)
        d = VAL * (n + 2)
        if num_digits(d) == n + 2:
            new_curr.add(d)
    curr = [VAL for VAL in new_curr]
    for VAL in curr:
        STRANGE_VALS.add(VAL)
STRANGE_VALS = sorted(STRANGE_VALS)

t = int(raw_input().strip())
for _ in range(t):
    L, R = map(int, raw_input().strip().split())
    l = 0
    while l < len(STRANGE_VALS) and L > STRANGE_VALS[l]:
        l += 1
    r = 0
    while r < len(STRANGE_VALS) and R >= STRANGE_VALS[r]:
        r += 1
    print r - l

File: practice/mathematics/number_theory/strange_numbers/strange_numbers.cpp
#include <cmath>
#include <vector>
#include <iostream>
#include <set>

using namespace std;

int num_digits(unsigned long long i_value)
{
    if (i_value == 0)
    {
        return 1;
    }
    return static_cast<int>(log10(i_value)) + 1;
}

vector<unsigned long long> strange_values()
{
    set<unsigned long long> strange;
    set<unsigned long long> curr;
    
    strange.insert(0);
    for (unsigned long long v = 1; v < 10; ++v)
    {
        strange.insert(v);
        curr.insert(v);
    }
    
    while (!curr.empty() && (*curr.begin() < 1000000000000000000ULL))
    {
        set<unsigned long long> new_curr;
        set<unsigned long long>::iterator it = curr.begin();
        for (; it != curr.end(); ++it)
        {
            int n = num_digits(*it);
            unsigned long long d = *it * n;
            if (d != *it && num_digits(d) == n)
            {
                new_curr.insert(d);
            }
            
            d = *it * (n + 1);
            if (num_digits(d) == n + 1)
            {
                new_curr.insert(d);
            }
            
            d = *it * (n + 2);
            if (num_digits(d) == n + 2)
            {
                new_curr.insert(d);
            }
        }
        
        curr = new_curr;
        it = curr.begin();
        for (; it != curr.end(); ++it)
        {
            strange.insert(*it);
        }
    }
    
    return vector<unsigned long long>(strange.begin(), strange.end());
}

int main()
{
    int t;
    unsigned long long L, R, l, r;
    
    vector<unsigned long long> strange = strange_values();
    
    cin >> t;
    while (t--)
    {
        cin >> L >> R;
        
        l = 0;
        while (l < strange.size() && L > strange[l])
        {
            l++;
        }
        
        r = 0;
        while (r < strange.size() && R >= strange[r])
        {
            r++;
        }
        
        cout << (r - l) << endl;
    }
    
    return 0;
}

